### PR TITLE
Several performance improvements

### PR DESCRIPTION
**Pull request by anonrig**
Before

```
BasicBench_AdaURL     175465 ns       175464 ns         3976 GHz=3.26934 cycle/byte=24.0253 cycles/url=746.974 instructions/byte=100.423 instructions/cycle=4.17988 instructions/ns=13.6654 instructions/url=3.12226k ns/url=228.478 speed=134.136M/s time/byte=7.45512ns time/url=231.788ns url/s=4.31428M/s
```

After

```
BasicBench_AdaURL     172468 ns       172467 ns         4089 GHz=3.24986 cycle/byte=23.5887 cycles/url=733.4 instructions/byte=99.9744 instructions/cycle=4.23823 instructions/ns=13.7737 instructions/url=3.10832k ns/url=225.671 speed=136.466M/s time/byte=7.32781ns time/url=227.83ns url/s=4.38924M/s
```

### CODE REVIEW

**Comment by anonrig** [comment]
I only see micro changes in the `wpt_bench` not on `bench`. Might be not worth the changes. @lemire. Feel free to close if you seem fit

File: src/helpers.cpp [hunk 1]
```diff
@@ -77,8 +77,11 @@ namespace ada::helpers {
     }
 
     // Remove path’s last item, if any.
-    if (!path.empty()) {
-      path.erase(path.rfind('/'));
+    for (long i = path.size() - 1; i >= 0; --i) {
```
Comment (lemire) [inline src/helpers.cpp:80]: It is not intuitive for me why this change might be faster. The rfind function might be well optimized.

File: src/unicode.cpp [hunk 2]
```diff
@@ -262,10 +262,8 @@ constexpr static uint8_t is_forbidden_domain_code_point_table[] = {
   }
 
   unsigned constexpr convert_hex_to_binary(const char c) noexcept {
-    // this code can be optimized.
-    if (c <= '9') { return c - '0'; }
-    char del = c >= 'a' ? 'a' : 'A';
-    return 10 + (c - del);
+    if (c <= '9') return c - '0';
```
Comment (lemire) [inline src/unicode.cpp:265]:
```suggestion
    if (c <= '9') return { c - '0'; }
```

Comment (lemire) [reply src/unicode.cpp:265]: If hex-to-binary is ever a bottleneck, there are much faster ways to deal with it. At this point in time, I suspect that it is rarely a bottleneck.

**Comment by lemire** [comment]
@anonrig I skeptical regarding one change... where you drop the `rfind` and replaced it by your own loop. It might be that it is faster... but this code...

```
if (!path.empty()) {
      path.erase(path.rfind('/'));
}
```

should be decent... It will trigger a function call (to rfind). I am not requesting that you change anything, but how sure are you that what you did resulted in better performance?

If the difference is too tiny to measure, I would simply revert to the simpler original code.

If you are convinced, merge away !!!
